state/application.go

```diff
@@ -1362,6 +1362,19 @@ func (a *Application) SetCharm(cfg SetCharmConfig) (err error) {
 				}}},
 			})
 		} else {
+			// Check if the new charm specifies a relation max limit
+			// that cannot be satisfied by the currently established
+			// relation count.
+			quotaErr := a.preUpgradeRelationLimitCheck(cfg.Charm)
```

Comment:
We should think about this long term, if this is the correct place for business logic?
At some point, we probably want to lift this up one level above state.

Reply:
Totally agree. This (and the other logic in the method) should be handled by a logic layer between the facade and state.